HDDS-8704. [Snapshot] Add CLI to list SnapDiff jobs

xBis7 opened the pull request:
## What changes were proposed in this pull request?

This patch adds a new CLI command for listing SnapshotDiff jobs for a bucket based on jobStatus. There is also the option to list all available jobs for the bucket.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8704

## How was this patch tested?

This patch was tested with new tests. Also, `snapshot-sh.robot` acceptance test was expanded to include listing snapshot diff jobs.

xBis7 commented:
@hemantk-12 @smengcl Can you take a look at this PR?

hemantk-12 commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/ObjectStore.java:
```diff
@@ -611,4 +612,21 @@ public SnapshotDiffResponse snapshotDiff(String volumeName,
+   * Get a list of the SnapshotDiff jobs for a bucket based on the JobStatus.
+   * @param volumeName Name of the volume to which the snapshotted bucket belong
+   * @param bucketName Name of the bucket to which the snapshots belong
+   * @param jobStatus JobStatus to be used to filter the snapshot diff jobs
```
Can you please add `listAll` too to this param list?

hemantk-12 commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/protocol/ClientProtocol.java:
```diff
@@ -1072,6 +1073,20 @@ SnapshotDiffResponse snapshotDiff(String volumeName, String bucketName,
+   * Get a list of the SnapshotDiff jobs for a bucket based on the JobStatus.
+   * @param volumeName Name of the volume to which the snapshotted bucket belong
+   * @param bucketName Name of the bucket to which the snapshots belong
+   * @param jobStatus JobStatus to be used to filter the snapshot diff jobs
```
Same as above.

hemantk-12 commented on hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:
```diff
@@ -135,6 +135,7 @@ enum Type {
   SnapshotPurge = 118;
   RecoverLease = 119;
   SetTimes = 120;
+  ListSnapshotDiffJob = 121;
```
nit: `ListSnapshotDiffJobs`.

hemantk-12 commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/SnapshotDiffJob.java:
```diff
@@ -15,14 +15,15 @@
  * limitations under the License.
  */
-package org.apache.hadoop.ozone.om.snapshot;
+package org.apache.hadoop.ozone.om.helpers;
```
Should we rename `SnapshotDiffJob` to `SnapshotDiffJobInfo` to be in alignment with other class?

xBis7 replied:
I wanted to make the class available to the client and keep diffs to a minimum. We will have to modify around 100 files to make this change.

hemantk-12 replied:
Yeah, it doesn't have to be part of this PR if we agree on it.

hemantk-12 commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/snapshot/SnapshotDiffResponse.java:
```diff
@@ -41,6 +41,15 @@ public JobStatusProto toProtobuf() {
       return JobStatus.valueOf(jobStatusProto.name());
     }
+
+    public static JobStatus getJobStatusFromString(String jobStatus) {
```
This helper function is not needed. Enum has [valueOf()](https://docs.oracle.com/javase/8/docs/api/java/lang/Enum.html#valueOf-java.lang.Class-java.lang.String-) which can be used directly. 

```
JobStatus.valueOf(jobStatus);
```

You just have to capitalized String value of [jobStatus](https://github.com/apache/ozone/pull/4824/files#diff-46f82d5fbac5dc7aae1fc363811fb88e6bc6c89a2aefafddad90de2f1ff3cfbdR401) here.

hemantk-12 commented on hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:
```diff
@@ -1709,6 +1726,13 @@ message SnapshotDiffRequest {
 }
 
+message ListSnapshotDiffJobRequest {
+  optional string volumeName = 1;
+  optional string bucketName = 2;
```
These two should be required.

hemantk-12 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java:
```diff
@@ -681,6 +686,14 @@ public SnapshotDiffResponse getSnapshotDiffReport(final String volume,
+                                                   final String bucketName,
+                                                   final String jobStatus,
+                                                   final boolean listAll) {
+    return snapshotDiffManager.getSnapshotDiffJobList(
```
I think we should validate that `volumeName` and `volumeName` exist to fail fast? May also add snapshot existed check.

xBis7 replied:
The only way I can think of to validate snapshots at this point, is to iterate the snapshot table. I think it's unnecessary since we are gonna get an empty list back, if there are no snapshots.

hemantk-12 replied:
I think you can use [path level SnapshotChain](https://github.com/apache/ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/SnapshotChainManager.java#L51).

xBis7 replied:
Done, I've also added an integration test to verify the error responses.

hemantk-12 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:
```diff
@@ -371,6 +378,35 @@ private Set<String> getSSTFileListForSnapshot(OmSnapshot snapshot,
+          // If provided job status is invalid,
+          // then all jobs on the table will be ignored.
+          // No need to check if getJobStatusFromString doesn't return null.
+          if (snapshotDiffJob.getStatus().equals(
```
It is better to use `==` for enum to be null safe.

https://stackoverflow.com/a/1750453

xBis7 replied:
> Alternative, could use [Objects.equals()](https://github.com/openjdk/jdk/blob/jdk8-b120/jdk/src/share/classes/java/util/Objects.java#L43-L60)?

That works as well.

hemantk-12 commented on hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:
```diff
@@ -1709,6 +1726,13 @@ message SnapshotDiffRequest {
   optional bool forceFullDiff = 7;
 }
 
+message ListSnapshotDiffJobRequest {
```
Should it be paginated? It doesn't have to be part of this PR.

xBis7 replied:
Other listing commands don't have pagination.

xBis7 replied:
okay, in that case we can do it in another PR.

hemantk-12 replied:
Please create a jira for that and resolve this comment.

xBis7 replied:
https://issues.apache.org/jira/browse/HDDS-8802

hemantk-12 commented on hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:
```diff
@@ -789,6 +794,18 @@ message SnapshotInfo {
  }
 
+message SnapshotDiffJobProto {
+  optional uint64 creationTime = 1;
+  optional string jobId = 2;
```
I'm not sure if we should return internal parameters like `volumeName`, `jobId` and `totalDiffEntries`.

xBis7 replied:
I will get rid of internal parameters.

xBis7 replied:
@hemantk-12 We can wrap `SnapshotDiffJob` with another class that has only the necessary parameters and then send that object as a response from server to client. In that case, `SnapshotDiffJob` will be moved back to its original package. What do you think?

hemantk-12 replied:
I see we return all the parameters in other list APIs [KeyInfo](https://github.com/apache/ozone/blob/9034434285f3877246b3ebc6a7513c082931b279/hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto#L1007) and [SnapshotInfo](https://github.com/apache/ozone/blob/9034434285f3877246b3ebc6a7513c082931b279/hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto#L777). You can ignore this comment for now.

hemantk-12 replied:
It is OK to return all the parameters here but we should have a client side `SnapshotDiffJob` similar to [OzoneKey](https://github.com/apache/ozone/blob/9c6cd4b93e0427fc12bbedae8156c746cbe6cea1/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneKey.java#L34) and [OzoneSnapshot](https://github.com/apache/ozone/blob/9c6cd4b93e0427fc12bbedae8156c746cbe6cea1/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneSnapshot.java#L27). And shows only information client parameters and hides the internal parameters.

xBis7 replied:
I'll make the changes.

xBis7 replied:
@hemantk-12 I've made the client changes and this is how it looks

```
bash-4.2$ ozone sh snapshot listDiff /vol1/bucket1 --all
[ {
  "volumeName" : "vol1",
  "bucketName" : "bucket1",
  "fromSnapshot" : "snap1",
  "toSnapshot" : "snap2",
  "jobStatus" : "DONE"
} ]
```

hemantk-12 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:
```diff
@@ -371,6 +378,35 @@ private Set<String> getSSTFileListForSnapshot(OmSnapshot snapshot,
+             snapDiffJobTable.iterator()) {
+      while (iterator.hasNext()) {
+        SnapshotDiffJob snapshotDiffJob = iterator.next().getValue();
+        if (snapshotDiffJob.getVolume().equals(volumeName) &&
```
nit: to be null safe.

```suggestion
        if (Objects.equals(snapshotDiffJob.getVolume(), volumeName) &&
            Objects.equals(snapshotDiffJob.getBucket(), bucketName)) {
```

xBis7 replied:
ah I get it. I'll keep it in mind.

hemantk-12 commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/snapshot/TestSnapshotDiffManager.java:
```diff
@@ -0,0 +1,328 @@
+  }
+
+  @Test
+  public void testListSnapshotDiffJobs()
```
nit: It would be nicer if you convert this test to [parameterized test](https://www.baeldung.com/parameterized-tests-junit-5). Repetitive code can be avoid with that.

Sample test: https://github.com/apache/ozone/blob/9034434285f3877246b3ebc6a7513c082931b279/hadoop-hdds/rocksdb-checkpoint-differ/src/test/java/org/apache/ozone/rocksdiff/TestRocksDBCheckpointDiffer.java#L263

xBis7 replied:
done.

hemantk-12 reviewed (commented):
Thanks for the patch @xBis7.

Left some comments.

xBis7 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/snapshot/ListSnapshotDiffHandler.java:
```diff
@@ -0,0 +1,84 @@
+    String volumeName = snapshotPath.getValue().getVolumeName();
+    String bucketName = snapshotPath.getValue().getBucketName();
+
+    if (Arrays.asList(STATUS_VALUES)
```
> Why not create STATUS_VALUES as a List itself?

I'll do that.

> Do we really need this check? Whenever we add a new status, we have to add it here as well.

This is to validate in the client whether the provided status is correct. We can skip this step and get an empty response from the server if the provided status doesn't exist.

xBis7 replied:
Removed it. Just like other listing commands it will just return an empty list.

hemantk-12 replied:
I agree that it is better to have client side validation. For this one, I'm not sure because client and server are not using the same code/interface.

xBis7 replied:
okay, let's do the validation on server side.

xBis7 commented:
@hemantk-12 I've addressed all your comments. The only ones left are about implementing pagination and renaming `SnapshotDiffJob` to `SnapshotDiffJobInfo`. These two can be addressed in separate PRs. 

I've also added a small test for validating error responses due to invalid volume name or bucket name or snapshot path.

hemantk-12 commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/snapshot/TestSnapshotDiffManager.java:
```diff
@@ -0,0 +1,333 @@
+    }
+  }
+
+  @ParameterizedTest
```
nit:
```suggestion
  private static Stream<Arguments> listSnapshotDiffJobsScenarios() {
    return Stream.of(
        Arguments.of("queued", false, false),
        Arguments.of("done", false, false),
        Arguments.of("in_progress", false, true),
        Arguments.of("invalid", false, false),
        Arguments.of("", false, false),
        Arguments.of("queued", true, true),
        Arguments.of("done", true, true),
        Arguments.of("in_progress", true, true),
        Arguments.of("invalid", true, true),
        Arguments.of("", true, true)
    );
  }

  @ParameterizedTest
  @MethodSource("listSnapshotDiffJobsScenarios")
  public void testListSnapshotDiffJobs(String jobStatus,
                                       boolean listAll,
                                       boolean containsJob) throws IOException {

    String fromSnapshotName = "snap-" + RandomStringUtils.randomNumeric(5);
    String toSnapshotName = "snap-" + RandomStringUtils.randomNumeric(5);
    String fromSnapshotId = UUID.randomUUID().toString();
    String toSnapshotId = UUID.randomUUID().toString();
    String diffJobKey = fromSnapshotId + DELIMITER + toSnapshotId;

    setUpKeysAndSnapshots(fromSnapshotName, toSnapshotName,
        fromSnapshotId, toSnapshotId);

    SnapshotDiffJob diffJob = snapDiffJobTable.get(diffJobKey);
    Assertions.assertNull(diffJob);

    // There are no jobs in the table, therefore
    // the response list should be empty.
    List<SnapshotDiffJob> jobList = snapshotDiffManager
        .getSnapshotDiffJobList(VOLUME, BUCKET, jobStatus, listAll);
    Assertions.assertTrue(jobList.isEmpty());

    
    SnapshotDiffResponse snapshotDiffResponse = snapshotDiffManager
        .getSnapshotDiffReport(VOLUME, BUCKET,
            fromSnapshotName, toSnapshotName,
            0, 0, false);
    
    Assertions.assertEquals(JobStatus.IN_PROGRESS,
        snapshotDiffResponse.getJobStatus());

    diffJob = snapDiffJobTable.get(diffJobKey);
    Assertions.assertNotNull(diffJob);
    Assertions.assertEquals(JobStatus.IN_PROGRESS,
        diffJob.getStatus());

    jobList = snapshotDiffManager
        .getSnapshotDiffJobList(VOLUME, BUCKET, jobStatus, listAll);

    if (containsJob) {
      Assertions.assertTrue(jobList.contains(diffJob));
    } else {
      Assertions.assertTrue(jobList.isEmpty());
    }
  }
```

On a side note, there is no need to add straight forward comments like  `// Submit a job.` or `// Response should be IN_PROGRESS.`

hemantk-12 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java:
```diff
@@ -767,6 +767,29 @@ public void testSnapDiffMultipleBuckets() throws Exception {
+    String errorMessage = "Provided volume name " + volume +
+        " or bucket name " + bucket + " doesn't exist";
+    // List SnapshotDiff jobs for non existent volume and bucket.
+    LambdaTestUtils.intercept(OMException.class,
```
nit: Use `Assert.assertThrows()` directly.

```suggestion
    OMException omException = assertThrows(OMException.class,
        () -> store.listSnapshotDiffJobs(volume, bucket, "", true));
   assertEquals(errorMessage, omException.getMessage());
```

hemantk-12 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java:
```diff
@@ -767,6 +767,29 @@ public void testSnapDiffMultipleBuckets() throws Exception {
+    ozVolume.createBucket(bucket);
+
+    // List SnapshotDiff jobs using a path that has no snapshots.
+    LambdaTestUtils.intercept(OMException.class,
```
```suggestion
    omException = assertThrows(OMException.class,
        () -> store.listSnapshotDiffJobs(volume, bucket, "", true));
    // List SnapshotDiff jobs using a path that has no snapshots.
   assertEquals("There are no snapshots", omException.getMessage());
```

hemantk-12 commented:
> @hemantk-12 I've addressed all your comments. The only ones left are about implementing pagination and renaming `SnapshotDiffJob` to `SnapshotDiffJobInfo`. These two can be addressed in separate PRs.
> 
> I've also added a small test for validating error responses due to invalid volume name or bucket name or snapshot path.

Thanks for adding integration test for invalid volume and bucket names.

hemantk-12 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java:
```diff
@@ -681,6 +686,37 @@ public SnapshotDiffResponse getSnapshotDiffReport(final String volume,
+        omMetadataManager.getSnapshotChainManager();
+    String snapshotPath = volumeName + OM_KEY_PREFIX + bucketName;
+    if (snapshotChainManager.getSnapshotChainPath(snapshotPath) == null) {
+      throw new IOException("There are no snapshots under path " +
```
Does it has to be an exception? Should we return empty response in that case?

xBis7 replied:
It's gonna return an empty response anyway. We can avoid iterating the snapshotDiffJob table by returning the empty response right there.

xBis7 commented:
@hemantk-12 I've addressed all your comments and also merged the branch into master and rewrote the tests to use the new `TestSnapshotDiffManager` setup.

hemantk-12 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/snapshot/ListSnapshotDiffHandler.java:
```diff
@@ -0,0 +1,74 @@
+/**
+ * ozone sh snapshot listSnapshotDiff.
+ */
+@CommandLine.Command(name = "listSnapshotDiff",
```
Similar to [HDDS-8789](https://issues.apache.org/jira/browse/HDDS-8789), I was thinking if we should simplify it and just call it listdiffJob. The command would be `ozone sh snapshot listdiffJob`. It would be great if we can do something like `ozone sh snapshot list diffJob` or `ozone sh snapshot ls diffJob`

xBis7 replied:
We can't have spaces, because the second word is considered part of the BucketURI. Another way to go about this, would be to make the command part of the diff command and have something like `ozone sh snapshot diff /vol/bucket list` or `ozone sh snapshot diff /vol/bucket ls`. The second approach might end up a bit complicated.

xBis7 replied:
@hemantk-12 That's how it looks.

```
bash-4.2$ ozone sh snapshot
Incomplete command
Usage: ozone sh snapshot [-hV] [COMMAND]
Snapshot specific operations
  -h, --help      Show this help message and exit.
  -V, --version   Print version information and exit.
Commands:
  create                                    Create a snapshot
  delete                                    Delete a snapshot
  list, ls                                  List snapshots for the buckets.
  snapshotDiff                              Get the differences between two
                                              snapshots
  listDiff, listDiffJob, lsDiff, lsDiffJob  List snapshotDiff jobs for a bucket.
```

hemantk-12 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/snapshot/ListSnapshotDiffHandler.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.List;
+
+/**
+ * ozone sh snapshot listSnapshotDiff.
```
nit: 
```suggestion
 * ozone sh snapshot listDiffJob.
 ```

smengcl commented:
Hi @xBis7 , looks like this PR has a conflict with HDDS-8490 (PR #4819 ) which is merged just now. Would you merge latest master into this PR to resolve the conflicts? Thx

prashantpogde reviewed (approved):
thanks @xBis7 .The changes look good to me.
